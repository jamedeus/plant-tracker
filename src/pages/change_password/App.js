import React, { useState, useMemo, useRef } from 'react';
import { useTheme } from 'src/context/ThemeContext';
import Navbar from 'src/components/Navbar';
import Cookies from 'js-cookie';
import clsx from 'clsx';

function App() {
    const { ToggleThemeOption } = useTheme();

    const formRef = useRef(null);
    const [oldPassword, setOldPassword] = useState('');
    const [newPassword1, setNewPassword1] = useState('');
    const [newPassword2, setNewPassword2] = useState('');
    const [oldPasswordIncorrect, setOldPasswordIncorrect] = useState(false);
    const [newPasswordError, setNewPasswordError] = useState(false);

    const submit = async (e) => {
        e.preventDefault();
        // Post old and new password to backend
        const response = await fetch("/accounts/change_password/submit/", {
            method: "POST",
            headers: {
                "Content-Type": "application/x-www-form-urlencoded",
                "X-CSRFToken": Cookies.get('csrftoken')
            },
            credentials: "include",
            body: new URLSearchParams({
                old_password: oldPassword,
                new_password1: newPassword1,
                new_password2: newPassword2,
            }),
        });
        // Redirect to overview if logged in successfully
        if (response.ok) {
            window.location.href = '/';
        // Show correct error message if login failed
        } else {
            const errors = await response.json();
            if (errors.errors.old_password) {
                setOldPasswordIncorrect(true);
            }
            if (errors.errors.new_password2) {
                setNewPasswordError(errors.errors.new_password2[0]);
            }
        }
    };

    const DropdownMenuOptions = useMemo(() => (
        <>
            <li><a onClick={() => window.location.href = "/"}>
                Overview
            </a></li>
            <ToggleThemeOption />
        </>
    ), [ToggleThemeOption]);

    return (
        <div className="container flex flex-col h-screen mx-auto items-center">
            <Navbar
                menuOptions={DropdownMenuOptions}
                title="Change Password"
            />

            <form ref={formRef} className="flex flex-col gap-4 mt-[15vh]">
                <label className="form-control w-full">
                    <div className="label">
                        <span className="label-text">Old password</span>
                    </div>
                    <input
                        name="old_password"
                        type="password"
                        className={clsx(
                            "input w-full input-bordered",
                            oldPasswordIncorrect && "input-error"
                        )}
                        value={oldPassword}
                        onInput={() => setOldPasswordIncorrect(false)}
                        onChange={(e) => setOldPassword(e.target.value)}
                    />
                </label>
                {oldPasswordIncorrect && (
                    <span className="text-error">
                        Old password incorrect
                    </span>
                )}
                <label className="form-control w-full relative">
                    <div className="label">
                        <span className="label-text">New password</span>
                    </div>
                    <input
                        name="new_password1"
                        type="password"
                        className={clsx(
                            "input w-full input-bordered",
                            newPasswordError && "input-error"
                        )}
                        value={newPassword1}
                        onChange={(e) => setNewPassword1(e.target.value)}
                    />
                </label>
                <label className="form-control w-full">
                    <div className="label">
                        <span className="label-text">Confirm new password</span>
                    </div>
                    <input
                        name="new_password2"
                        type="password"
                        className={clsx(
                            "input w-full input-bordered",
                            newPasswordError && "input-error"
                        )}
                        value={newPassword2}
                        onInput={() => setNewPasswordError(false)}
                        onChange={(e) => setNewPassword2(e.target.value)}
                    />
                </label>
                {newPasswordError && (
                    <span className="text-error">
                        {newPasswordError}
                    </span>
                )}

                <button
                    className="btn btn-success mt-6"
                    onClick={(e) => submit(e)}
                >
                    Change password
                </button>
            </form>
        </div>
    );
}

export default App;
