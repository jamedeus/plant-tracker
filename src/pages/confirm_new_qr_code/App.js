import React, { useEffect, useMemo } from 'react';
import PropTypes from 'prop-types';
import { sendPostRequest, parseDomContext } from 'src/util';
import Navbar from 'src/components/Navbar';
import NavbarDropdownOptions from 'src/components/NavbarDropdownOptions';
import PlantDetails from 'src/components/PlantDetails.js';
import GroupDetails from 'src/components/GroupDetails.js';
import { openErrorModal } from 'src/components/ErrorModal';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faXmark, faCheck } from '@fortawesome/free-solid-svg-icons';

const Layout = ({ type, thumbnailUrl, handleConfirm }) => {
    return (
        <div className="flex flex-col gap-8 text-center my-auto">
            <p className="text-lg font-bold">
                Is this the new QR code for your {type}?
            </p>
            {thumbnailUrl && (
                <div className="mx-auto p-4 bg-base-200 rounded-3xl">
                    <img
                        className="max-h-[50vh] rounded-xl object-contain"
                        src={thumbnailUrl}
                    />
                </div>
            )}

            {/* Confirm/cancel buttons */}
            <div className="flex gap-4 mx-auto mb-8">
                <button
                    className="btn btn-error btn-square text-white"
                    onClick={() => window.location.href = "/"}
                    title="Don't change QR code"
                >
                    <FontAwesomeIcon className="size-6" icon={faXmark} />
                </button>
                <button
                    className="btn btn-success btn-square text-white"
                    onClick={handleConfirm}
                    title="Change QR code"
                >
                    <FontAwesomeIcon className="size-6" icon={faCheck} />
                </button>
            </div>
        </div>
    );
};

Layout.propTypes = {
    type: PropTypes.oneOf([
        'plant',
        'group'
    ]).isRequired,
    thumbnailUrl: PropTypes.string,
    handleConfirm: PropTypes.func.isRequired
};

function App() {
    // Load UUIDs and plant/group state from django template context
    const type = parseDomContext("type");
    const instance = parseDomContext("instance");
    const newUuid = parseDomContext("new_uuid");

    // Reload if user navigates to page by pressing back button (change QR code
    // request likely expired, replace with register or manage page)
    useEffect(() => {
        const handleBackButton = async (event) => {
            if (event.persisted) {
                window.location.reload();
            }
        };

        // Add listener on mount, remove on unmount
        window.addEventListener('pageshow', handleBackButton);
        return () => {
            window.removeEventListener('pageshow', handleBackButton);
        };
    }, []);

    const handleConfirm = async () => {
        const response = await sendPostRequest(
            '/change_uuid',
            {
                uuid: instance.uuid,
                new_id: newUuid
            }
        );
        // Reload page if changed successfully
        if (response.ok) {
            window.location.reload();
        } else {
            const error = await response.json();
            openErrorModal(JSON.stringify(error));
        }
    };

    // Plant/group details card shown when title is clicked
    const DetailsDropdown = useMemo(() => (
        <div className="details-card">
            <div className="card-body">
                {type === 'plant' ? (
                    <PlantDetails
                        species={instance.species}
                        pot_size={instance.pot_size}
                        description={instance.description}
                    />
                ) : (
                    <GroupDetails
                        location={instance.location}
                        description={instance.description}
                    />
                )}
            </div>
        </div>
    ), []);

    return (
        <div className="container flex flex-col mx-auto h-screen">
            <Navbar
                menuOptions={<NavbarDropdownOptions />}
                title={instance.display_name}
                titleOptions={DetailsDropdown}
            />
            <Layout
                type={type}
                thumbnailUrl={instance.thumbnail}
                handleConfirm={handleConfirm}
            />
        </div>
    );
}

export default App;
