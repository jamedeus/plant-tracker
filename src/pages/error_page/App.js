import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import Navbar from 'src/components/Navbar';
import NavbarDropdownOptions from 'src/components/NavbarDropdownOptions';

function App({ errorMessage }) {
    return (
        <div
            className="container flex flex-col full-screen mx-auto items-center gap-16"
            data-testid="error-page"
        >
            <Navbar
                menuOptions={<NavbarDropdownOptions />}
                title="Error"
                showScanButton={false}
            />

            <div className="flex flex-col text-center text-lg mt-[15vh]">
                {errorMessage}
            </div>

            <Link to="/" className='btn btn-accent' discover="none">
                Go to Overview
            </Link>
        </div>
    );
}

export default App;

App.propTypes = {
    errorMessage: PropTypes.string.isRequired
};
