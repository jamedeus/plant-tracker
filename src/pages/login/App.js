import React, { useState, useMemo, useRef } from 'react';
import { useTheme } from 'src/context/ThemeContext';
import Navbar from 'src/components/Navbar';
import Cookies from 'js-cookie';
import clsx from 'clsx';

function App() {
    const { ToggleThemeOption } = useTheme();

    const formRef = useRef(null);
    const [username, setUsername] = useState('');
    const [password, setPassword] = useState('');
    const [showError, setShowError] = useState(false);

    const submit = async (e) => {
        e.preventDefault();
        // Post FormData to backend
        const formData = new FormData(formRef.current);
        const response = await fetch('/accounts/login/', {
            method: 'POST',
            body: formData,
            headers: {
                'Accept': 'application/json, text/plain, */*',
                "X-CSRFToken": Cookies.get('csrftoken')
            }
        });
        // Redirect if logged in successfully
        if (response.ok) {
            // Redirect to url in querystring if present
            const params = new URL(window.location.href).searchParams;
            if (params.get('next')) {
                window.location.href = params.get('next');
            // Redirect to overview if no querystring
            } else {
                window.location.href = '/';
            }
        // Show error text if login failed
        } else {
            setShowError(true);
        }
    };

    const DropdownMenuOptions = useMemo(() => (
        <>
            <li><a onClick={() => window.location.href = "/"}>
                Overview
            </a></li>
            <ToggleThemeOption />
        </>
    ), [ToggleThemeOption]);

    return (
        <div className="container flex flex-col h-screen mx-auto items-center">
            <Navbar
                menuOptions={DropdownMenuOptions}
                title="Login"
            />

            <form ref={formRef} className="flex flex-col gap-4 mt-[15vh]">
                <label className="form-control w-full">
                    <div className="label">
                        <span className="label-text">Username</span>
                    </div>
                    <input
                        name="username"
                        type="text"
                        autoCapitalize="off"
                        className={clsx(
                            "input w-full input-bordered",
                            showError && "input-error"
                        )}
                        value={username}
                        onInput={() => setShowError(false)}
                        onChange={e => setUsername(e.target.value)}
                    />
                </label>
                <label className="form-control w-full relative">
                    <div className="label">
                        <span className="label-text">Password</span>
                    </div>
                    <input
                        name="password"
                        type="password"
                        className={clsx(
                            "input w-full input-bordered",
                            showError && "input-error"
                        )}
                        value={password}
                        onInput={() => setShowError(false)}
                        onChange={e => setPassword(e.target.value)}
                    />
                </label>

                <button
                    className="btn btn-success mt-6"
                    onClick={(e) => submit(e)}
                    disabled={username.length < 1 || password.length < 1}
                >
                    Login
                </button>

                {showError && (
                    <span className="text-error text-center">
                        Invalid username or password
                    </span>
                )}
            </form>
        </div>
    );
}

export default App;
