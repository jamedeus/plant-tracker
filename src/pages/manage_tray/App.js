import React, { useState, useRef } from 'react';
import PropTypes from 'prop-types';
import { sendPostRequest, parseDomContext, localToUTC, pastTense } from 'src/util';
import EditModal from 'src/components/EditModal';
import TrayDetailsForm from 'src/forms/TrayDetailsForm';
import Navbar from 'src/components/Navbar';
import PlantCard from 'src/components/PlantCard';
import DatetimeInput from 'src/components/DatetimeInput';
import FilterColumn from 'src/components/FilterColumn';
import { useToast } from 'src/context/ToastContext';
import { useTheme } from 'src/context/ThemeContext';
import DetailsCard from 'src/components/DetailsCard';
import TrayDetails from 'src/components/TrayDetails';
import AddPlantsModal, { openAddPlantsModal } from './AddPlantsModal';
import RemovePlantsModal, { openRemovePlantsModal } from './RemovePlantsModal';
import { useErrorModal } from 'src/context/ErrorModalContext';

function App() {
    // Load context set by django template
    const [tray, setTray] = useState(() => {
        return parseDomContext("tray");
    });
    const [plantDetails, setPlantDetails] = useState(() => {
        return parseDomContext("details");
    });
    // Contains list of objects with name and uuid of every plant in database
    // DO NOT mutate (used to generate add/remove plant menu options)
    const options = parseDomContext("options");

    // Create state to track whether selecting plants from list
    const [selectingPlants, setSelectingPlants] = useState(false);

    // Track which plants are selected (after clicking manage button)
    const selectedPlants = useRef([]);

    // Track plants column open/close state between re-renders
    const plantsOpenRef = useRef(true);

    // Ref to access timestamp input used by water all/fertilize all
    const addEventAllTimeInput = useRef(null);

    // Create ref to access edit details form
    const editDetailsRef = useRef(null);

    // Get hooks to show toast message, error modal
    const { showToast } = useToast();
    const { showErrorModal } = useErrorModal();

    // Get toggle theme option from context
    const { ToggleThemeOption } = useTheme();

    const submitEditModal = async () => {
        const payload = Object.fromEntries(
            new FormData(editDetailsRef.current)
        );
        payload["tray_id"] = tray.uuid;
        console.log(payload);

        const response = await sendPostRequest('/edit_tray', payload);
        if (response.ok) {
            // Update plant state with new values from response
            const data = await response.json();
            setTray({...tray, ...data});
        } else {
            const error = await response.json();
            showErrorModal(JSON.stringify(error));
        }
    };

    // Handler for "Water All" and "Fertilize All" buttons
    const addEventAll = async (eventType) => {
        const timestamp = localToUTC(addEventAllTimeInput.current.value);
        // Post eventType, all plant UUIDs, and timestamp to backend endpoint
        await bulkAddPlantEvents(
            eventType,
            plantDetails.map(plant => plant.uuid),
            timestamp
        );
    };

    // Creates event with specified type and timestamp for every plant in
    // selectedIds (array of UUIDs)
    const bulkAddPlantEvents = async (eventType, selectedIds, timestamp) => {
        const payload = {
            plants: selectedIds,
            event_type: eventType,
            timestamp: timestamp
        };
        const response = await sendPostRequest('/bulk_add_plant_events', payload);
        if (response.ok) {
            showToast(`All plants ${pastTense(eventType)}!`, 'blue', 5000);
            const data = await response.json();
            updatePlantTimestamps(data.plants, timestamp, eventType);
        } else {
            const error = await response.json();
            showErrorModal(JSON.stringify(error));
        }
    };

    // Map eventType taken by bulk_add_plant_events to the plantDetails state
    // key that should be updated when an event is successfully created
    const eventTypeMap = {
        water: "last_watered",
        fertilize: "last_fertilized"
    };

    // Called by bulkAddPlantEvents to update plant water/fertilize timestamps
    const updatePlantTimestamps = (updatedPlants, timestamp, eventType) => {
        let newPlantDetails = [];
        plantDetails.forEach(plant => {
            // Overwrite last_watered if UUID in JSON response
            if (updatedPlants.includes(plant.uuid)) {
                plant[eventTypeMap[eventType]] = timestamp;
            }
            newPlantDetails.push(plant);
        });
        setPlantDetails(newPlantDetails);
    };

    // Buttons used to add bulk events to plants in tray
    const PlantEventButtons = ({editing, setEditing}) => {
        const addEventTimeInput = useRef(null);

        // Handler for water button (only used in this case scope)
        const water = async () => {
            const timestamp = localToUTC(addEventTimeInput.current.value);
            await bulkAddPlantEvents('water', selectedPlants.current, timestamp);
            setEditing(false);
        };

        // Handler for fertilize button (only used in this case scope)
        const fertilize = async () => {
            const timestamp = localToUTC(addEventTimeInput.current.value);
            await bulkAddPlantEvents('fertilize', selectedPlants.current, timestamp);
            setEditing(false);
        };

        switch(editing) {
            case(true):
                return (
                    <>
                        <div
                            className="flex mx-auto mb-4"
                            data-testid="addEventTimeInput"
                        >
                            <DatetimeInput inputRef={addEventTimeInput} />
                        </div>
                        <div className="flex">
                            <button
                                className="btn btn-outline mx-auto"
                                onClick={() => setEditing(false)}
                            >
                                Cancel
                            </button>
                            <button
                                className="btn btn-outline btn-info mx-auto"
                                onClick={water}
                            >
                                Water
                            </button>
                            <button
                                className="btn btn-outline btn-success mx-auto"
                                onClick={fertilize}
                            >
                                Fertilize
                            </button>
                        </div>
                    </>
                );
            case(false):
                return (
                    <div className="flex">
                        <button
                            className="btn btn-outline mx-auto"
                            onClick={() => setEditing(true)}>
                            Manage
                        </button>
                    </div>
                );
        }
    };

    PlantEventButtons.propTypes = {
        editing: PropTypes.bool,
        setEditing: PropTypes.func
    };

    return (
        <div className="container flex flex-col mx-auto mb-8">
            <Navbar
                menuOptions={
                    <>
                        <li><a onClick={() => window.location.href = "/"}>
                            Overview
                        </a></li>
                        <li><a onClick={openAddPlantsModal}>
                            Add plants
                        </a></li>
                        <li><a onClick={openRemovePlantsModal}>
                            Remove plants
                        </a></li>
                        <ToggleThemeOption />
                    </>
                }
                title={tray.display_name}
                titleOptions={
                    <DetailsCard>
                        <TrayDetails
                            location={tray.location}
                            description={tray.description}
                        />
                    </DetailsCard>
                }
            />

            <DatetimeInput inputRef={addEventAllTimeInput} />
            <div className="flex mx-auto mb-8">
                <button
                    className="btn btn-info m-2"
                    onClick={() => addEventAll('water')}
                >
                    Water All
                </button>
                <button
                    className="btn btn-success m-2"
                    onClick={() => addEventAll('fertilize')}
                >
                    Fertilize All
                </button>
            </div>

            <FilterColumn
                title="Plants"
                contents={plantDetails}
                CardComponent={PlantCard}
                editing={selectingPlants}
                selected={selectedPlants}
                openRef={plantsOpenRef}
            >
                <PlantEventButtons
                    editing={selectingPlants}
                    setEditing={setSelectingPlants}
                />
            </FilterColumn>

            <EditModal title="Edit Details" onSubmit={submitEditModal}>
                <TrayDetailsForm
                    formRef={editDetailsRef}
                    name={tray.name}
                    location={tray.location}
                    description={tray.description}
                />
            </EditModal>

            <AddPlantsModal
                trayID={tray.uuid}
                options={options}
                plantDetails={plantDetails}
                setPlantDetails={setPlantDetails}
            />

            <RemovePlantsModal
                trayID={tray.uuid}
                plantDetails={plantDetails}
                setPlantDetails={setPlantDetails}
            />
        </div>
    );
}

export default App;
