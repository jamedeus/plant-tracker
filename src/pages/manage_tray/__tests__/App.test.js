import renderer from 'react-test-renderer';
import { DateTime } from 'src/testUtils/luxonMock';
import createMockContext from 'src/testUtils/createMockContext';
import App from '../App';
import { ToastProvider } from 'src/ToastContext';

// Simulated django context, parsed into state object
const mockContext = {
    "tray": {
        "uuid": "0640ec3b-1bed-4b15-a078-d6e7ec66be14",
        "name": "Test tray",
        "display_name": "Test tray",
        "location": "Middle shelf",
        "description": null
    },
    "details": [
        {
            "name": "Test Plant",
            "uuid": "0640ec3b-1bed-4b15-a078-d6e7ec66be12",
            "species": "Calathea",
            "description": "This is a plant with a long description with",
            "pot_size": 4,
            "last_watered": "2024-03-01T05:45:44+00:00",
            "last_fertilized": "2024-03-01T05:45:44+00:00"
        },
        {
            "name": "Unnamed Spider Plant",
            "uuid": "19f65fa0-1c75-4cba-b590-0c9b5b315fcc",
            "species": "Spider Plant",
            "description": null,
            "pot_size": 2,
            "last_watered": "2024-03-01T05:45:44+00:00",
            "last_fertilized": "2024-03-01T05:45:44+00:00"
        },
    ],
    "options": [
        {
            "uuid": "0640ec3b-1bed-4b15-a078-d6e7ec66be12",
            "name": "Test Plant"
        },
        {
            "uuid": "0640ec3b-1bed-4b15-a078-d6e7ec66be16",
            "name": "Another test plant"
        },
        {
            "uuid": "0640ec3b-1bed-4b15-a078-d6e7ec66be69",
            "name": "Third test plant"
        },
        {
            "uuid": "19f65fa0-1c75-4cba-b590-0c9b5b315fcc",
            "name": "Unnamed Spider Plant"
        }
    ]
}

describe('App', () => {
    // Mock long-supported features that jsdom somehow hasn't implemented yet
    beforeAll(() => {
        HTMLDialogElement.prototype.show = jest.fn();
        HTMLDialogElement.prototype.showModal = jest.fn();
        HTMLDialogElement.prototype.close = jest.fn();
    });

    // Setup: Create mock state objects
    beforeEach(() => {
        createMockContext('tray', mockContext.tray);
        createMockContext('details', mockContext.details);
        createMockContext('options', mockContext.options);
    });

    it('matches snapshot', () => {
        // Mock system time so relative times ("1 hour ago") don't change
        jest.useFakeTimers();
        jest.setSystemTime(new Date('2024-03-01T12:00:00Z'));

        const component = renderer.create(
            <ToastProvider>
                <App />
            </ToastProvider>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();

        // Reset mock
        jest.useRealTimers();
    });
});
