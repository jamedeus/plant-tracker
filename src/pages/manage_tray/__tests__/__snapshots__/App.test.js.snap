// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot 1`] = `
[
  <div
    className="container flex flex-col mx-auto mb-8"
  >
    <div
      className="navbar bg-base-100 mb-4 sticky top-0 z-[99]"
    >
      <div
        className="navbar-start"
      >
        <div
          className="dropdown"
        >
          <div
            className="btn btn-ghost btn-circle"
            role="button"
            tabIndex={0}
          >
            <svg
              className="h-5 w-5"
              fill="none"
              stroke="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M4 6h16M4 12h16M4 18h7"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
              />
            </svg>
          </div>
          <ul
            className="menu menu-md dropdown-content mt-3 z-[99]
                                    p-2 shadow bg-base-300 rounded-box w-52"
            tabIndex={0}
          >
            <li>
              <a
                onClick={[Function]}
              >
                Overview
              </a>
            </li>
            <li>
              <a
                onClick={[Function]}
              >
                Add plants
              </a>
            </li>
            <li>
              <a
                onClick={[Function]}
              >
                Remove plants
              </a>
            </li>
          </ul>
        </div>
      </div>
      <div
        className="navbar-center"
      >
        <div
          className="dropdown dropdown-center"
        >
          <a
            className="btn btn-ghost text-3xl"
            role="button"
            tabIndex={0}
          >
            Test tray
          </a>
          <div
            className="dropdown-content z-[1] flex"
            tabIndex={0}
          >
            <div
              className="card card-compact w-72 p-2 mt-2 mx-auto
                        shadow bg-neutral text-neutral-content"
            >
              <div
                className="card-body"
              >
                <p
                  className="flex"
                >
                  <span
                    className="font-semibold"
                  >
                    Location:
                  </span>
                  <span
                    className="ml-auto"
                  >
                    Middle shelf
                  </span>
                </p>
                <div
                  className="hidden"
                >
                  <p
                    className="font-semibold mt-3"
                  >
                    Description:
                  </p>
                  <p
                    className="text-sm line-clamp-6"
                  />
                </div>
                <button
                  className="btn btn-sm mt-4"
                  onClick={[Function]}
                >
                  Edit
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="navbar-end"
      />
    </div>
    <input
      className="input input-bordered mx-auto my-2"
      defaultValue="2024-03-01T12:00:00"
      id="addEventAllTime"
      step="1"
      type="datetime-local"
    />
    <div
      className="flex mx-auto mb-8"
    >
      <button
        className="btn btn-info m-2"
        onClick={[Function]}
      >
        Water All
      </button>
      <button
        className="btn btn-success m-2"
        onClick={[Function]}
      >
        Fertilize All
      </button>
    </div>
    <div
      className="collapse bg-base-200 w-96 px-4 mx-auto max-w-90 md:max-w-full"
    >
      <input
        defaultChecked={true}
        onChange={[Function]}
        type="checkbox"
      />
      <div
        className="collapse-title text-xl font-medium text-center"
      >
        Plants (2)
      </div>
      <div
        className="collapse-content"
      >
        <div
          className="px-4 mb-4 relative"
        >
          <input
            className="input input-bordered w-full text-center pr-10"
            onChange={[Function]}
            placeholder="filter"
            style={
              {
                "textIndent": "1.625rem",
              }
            }
            type="text"
            value=""
          />
          <button
            className="btn-close absolute right-6 top-2
                        opacity-0 -z-10"
            onClick={[Function]}
          >
            <svg
              aria-hidden="true"
              className="w-8 h-8"
              data-slot="icon"
              fill="currentColor"
              viewBox="0 0 16 16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.28 4.22a.75.75 0 0 0-1.06 1.06L6.94 8l-2.72 2.72a.75.75 0 1 0 1.06 1.06L8 9.06l2.72 2.72a.75.75 0 1 0 1.06-1.06L9.06 8l2.72-2.72a.75.75 0 0 0-1.06-1.06L8 6.94 5.28 4.22Z"
              />
            </svg>
          </button>
        </div>
        <div
          className="mb-4"
        >
          <div
            className="card bg-neutral text-neutral-content mx-auto relative"
            onClick={[Function]}
          >
            <div
              className="card-body text-center"
            >
              <h2
                className="card-title mx-auto pr-8 indent-8 line-clamp-1"
              >
                Test Plant
              </h2>
              Watered 3 hours ago
            </div>
            <button
              className="btn-close absolute right-2 top-8 z-40"
              onClick={[Function]}
              role="button"
              tabIndex={0}
            >
              <svg
                aria-hidden="true"
                className="w-8 h-8"
                data-slot="icon"
                fill="currentColor"
                viewBox="0 0 16 16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clipRule="evenodd"
                  d="M4.22 6.22a.75.75 0 0 1 1.06 0L8 8.94l2.72-2.72a.75.75 0 1 1 1.06 1.06l-3.25 3.25a.75.75 0 0 1-1.06 0L4.22 7.28a.75.75 0 0 1 0-1.06Z"
                  fillRule="evenodd"
                />
              </svg>
            </button>
            <div
              className="collapse bg-neutral"
            >
              <input
                className="hidden"
                defaultChecked={false}
                onChange={[Function]}
                type="checkbox"
              />
              <div
                className="collapse-content px-16"
              >
                <p
                  className="flex"
                >
                  <span
                    className="font-semibold"
                  >
                    Species:
                  </span>
                  <span
                    className="ml-auto"
                  >
                    Calathea
                  </span>
                </p>
                <p
                  className="flex"
                >
                  <span
                    className="font-semibold"
                  >
                    Pot size:
                  </span>
                  <span
                    className="ml-auto"
                  >
                    4
                  </span>
                </p>
                <div
                  className="text-center"
                >
                  <p
                    className="font-semibold mt-3"
                  >
                    Description:
                  </p>
                  <p
                    className="text-sm line-clamp-6"
                  >
                    This is a plant with a long description with
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="mb-4"
        >
          <div
            className="card bg-neutral text-neutral-content mx-auto relative"
            onClick={[Function]}
          >
            <div
              className="card-body text-center"
            >
              <h2
                className="card-title mx-auto pr-8 indent-8 line-clamp-1"
              >
                Unnamed Spider Plant
              </h2>
              Watered 3 hours ago
            </div>
            <button
              className="btn-close absolute right-2 top-8 z-40"
              onClick={[Function]}
              role="button"
              tabIndex={0}
            >
              <svg
                aria-hidden="true"
                className="w-8 h-8"
                data-slot="icon"
                fill="currentColor"
                viewBox="0 0 16 16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clipRule="evenodd"
                  d="M4.22 6.22a.75.75 0 0 1 1.06 0L8 8.94l2.72-2.72a.75.75 0 1 1 1.06 1.06l-3.25 3.25a.75.75 0 0 1-1.06 0L4.22 7.28a.75.75 0 0 1 0-1.06Z"
                  fillRule="evenodd"
                />
              </svg>
            </button>
            <div
              className="collapse bg-neutral"
            >
              <input
                className="hidden"
                defaultChecked={false}
                onChange={[Function]}
                type="checkbox"
              />
              <div
                className="collapse-content px-16"
              >
                <p
                  className="flex"
                >
                  <span
                    className="font-semibold"
                  >
                    Species:
                  </span>
                  <span
                    className="ml-auto"
                  >
                    Spider Plant
                  </span>
                </p>
                <p
                  className="flex"
                >
                  <span
                    className="font-semibold"
                  >
                    Pot size:
                  </span>
                  <span
                    className="ml-auto"
                  >
                    2
                  </span>
                </p>
                <div
                  className="hidden"
                >
                  <p
                    className="font-semibold mt-3"
                  >
                    Description:
                  </p>
                  <p
                    className="text-sm line-clamp-6"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="flex"
        >
          <button
            className="btn btn-outline mx-auto"
            onClick={[Function]}
          >
            Manage
          </button>
        </div>
      </div>
    </div>
    <dialog
      className="modal"
      id="editModal"
    >
      <div
        className="modal-box text-center flex flex-col pt-4"
      >
        <form
          method="dialog"
        >
          <button
            className="btn-close absolute right-4 top-4"
          >
            <svg
              aria-hidden="true"
              className="w-8 h-8"
              data-slot="icon"
              fill="currentColor"
              viewBox="0 0 16 16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.28 4.22a.75.75 0 0 0-1.06 1.06L6.94 8l-2.72 2.72a.75.75 0 1 0 1.06 1.06L8 9.06l2.72 2.72a.75.75 0 1 0 1.06-1.06L9.06 8l2.72-2.72a.75.75 0 0 0-1.06-1.06L8 6.94 5.28 4.22Z"
              />
            </svg>
          </button>
        </form>
        <h3
          className="font-bold text-lg"
        >
          Edit Details
        </h3>
        <form
          id="trayDetails"
        >
          <label
            className="form-control w-full"
          >
            <div
              className="label"
            >
              <span
                className="label-text-alt"
              >
                Tray name
              </span>
            </div>
            <input
              className="input w-full input-bordered"
              defaultValue="Test tray"
              name="name"
              type="text"
            />
          </label>
          <label
            className="form-control w-full"
          >
            <div
              className="label"
            >
              <span
                className="label-text-alt"
              >
                Tray location
              </span>
            </div>
            <input
              className="input w-full input-bordered"
              defaultValue="Middle shelf"
              name="location"
              type="text"
            />
          </label>
          <label
            className="form-control w-full"
          >
            <div
              className="label"
            >
              <span
                className="label-text-alt"
              >
                Description
              </span>
            </div>
            <textarea
              className="textarea textarea-bordered"
              defaultValue={null}
              name="description"
            />
          </label>
        </form>
        <div
          className="modal-action mx-auto"
        >
          <form
            method="dialog"
          >
            <button
              className="btn mr-2"
            >
              Cancel
            </button>
            <button
              className="btn ml-2"
              onClick={[Function]}
            >
              Edit
            </button>
          </form>
        </div>
      </div>
      <form
        className="modal-backdrop"
        method="dialog"
      >
        <button>
          close
        </button>
      </form>
    </dialog>
    <dialog
      className="modal"
      id="managePlantsModal"
    >
      <div
        className="modal-box text-center flex flex-col pt-4"
      >
        <form
          method="dialog"
        >
          <button
            className="btn-close absolute right-4 top-4"
          >
            <svg
              aria-hidden="true"
              className="w-8 h-8"
              data-slot="icon"
              fill="currentColor"
              viewBox="0 0 16 16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.28 4.22a.75.75 0 0 0-1.06 1.06L6.94 8l-2.72 2.72a.75.75 0 1 0 1.06 1.06L8 9.06l2.72 2.72a.75.75 0 1 0 1.06-1.06L9.06 8l2.72-2.72a.75.75 0 0 0-1.06-1.06L8 6.94 5.28 4.22Z"
              />
            </svg>
          </button>
        </form>
      </div>
      <form
        className="modal-backdrop"
        method="dialog"
      >
        <button>
          close
        </button>
      </form>
    </dialog>
  </div>,
  <div
    className="toast toast-center opacity-0 transition-all duration-500"
  >
    <div
      className="alert alert-info gap-0"
    >
      <span />
    </div>
  </div>,
]
`;
